text/0000-portals.md

```diff
@@ -0,0 +1,110 @@
+---
+title: portals proposal
+status: DRAFTED
+created_at: 2019-12-23
+updated_at: 2019-12-23
+pr:
+---
+
+# Short Title
+Out of tree DOM insertion of elements via portals.
+
+## Summary
+Introduce a mechanism in LWC for inserting and managing an element outside of its natural dom tree.
+
+## Motivation
+
+Position of a node in the DOM determines its stacking context with the component having no 
+ability to change it other than create a new stacking context. 
+
+This becomes problematic when one wants to present a popup/tooltip/dialog/dropdown/notification that needs to appear on top of other content on the screen, possibly limited by a parent with `overflow: hidden`, or a higher `z-index`. A first instinct for going around `overflow: hidden` is to use `position: fixed`, but it's also bound by the stacking context, and may not work as expected.
+
+Having a custom element be inserted out of its natural dom tree hierarchy would enable one to have greater control over the stacking context, allowing a predictalbe popup and modal implementation (including faux-modals that are scoped to tabs and are not per-se global).
+
+Side-note: Modal scoping means that a modal would be tied to a tab (the backdrop only appears inside the tab, and the modal is centered against the tab content) and only blocks the content of the tab, and nothing outside of it (focus trapping still works as with a true modal, but there is a key that allows one to escape the modal). When using Lightning Console in Salesforce, all modals by default are scoped, so this is a very important use-case to capture.
+
+## Basic example
+
+The proposal suggests one new API -- a way to define the destination of an element (position in the DOM where the component would be rendered in). The api in question is a new attribute -- `lwc:portal` that accepts a string that should match the global document id of the element that would contain the rendered element.
```

Comment:
Instead of `id` which will get mangled, we should use another directive like `lwc:portal:id` or something along those lines.

Reply:
@ekashida that's what I had initially (can see it in the commit history), this indeed is a problem, so I may need to go back to the original version with the portal definition

Reply:
@sfdciuie thanks for pointing out the comment history. Regarding `lightning-portal`, I guess we should avoid component dependencies for framework features.

After reading that discussion and thinking more about it, I agree there are no issues if the portal root lives outside of any component as a direct descendant of `document.body`. Also, if the need to bypass id-mangling ever arises, we can just add a new template directive to do so.